bundles/tools.mdsd.characteristics.edit/src/tools/mdsd/characteristics/edit/support/CharacteristicsContextItemProviderDecorator.java

```diff
@@ -0,0 +1,143 @@
+		@Override
+		public void notifyChanged(Notification notification) {
+			if (BindingPackage.eINSTANCE.getCharacterizationContext_Bindings().equals(notification.getFeature())) {
+				characteristicBasedDescriptors.remove(notification.getNotifier());
+			} else if (BindingPackage.eINSTANCE.getCharacteristicBinding().isInstance(notification.getNotifier())) {
+				EObject ctx = ((CharacteristicBinding)notification.getNotifier()).eContainer();
+				characteristicBasedDescriptors.remove(ctx);
+			}
+		} 
+	}
+
+	WeakHashMap<CharacterizationContext, List<IItemPropertyDescriptor>> characteristicBasedDescriptors = new WeakHashMap<>();
+
+	public CharacteristicsContextItemProviderDecorator(AdapterFactory adapterFactory) {
+		super(adapterFactory);
+	}
+
+	@Override
+	public List<IItemPropertyDescriptor> getPropertyDescriptors(Object object) {
+		List<IItemPropertyDescriptor> result = super.getPropertyDescriptors(object);
+		if ((object instanceof CharacterizationContext)) {
+			CharacterizationContext ctx = (CharacterizationContext) object;
+			List<IItemPropertyDescriptor> characteristicBased = getCharacteristicBasedDescriptors(ctx);
+			result.addAll(characteristicBased);
+		}
+		return result;
+	}
+
+	@Override
+	public IItemPropertyDescriptor getPropertyDescriptor(Object object, Object propertyId) {
+		if ((object instanceof CharacterizationContext)) {
+			CharacterizationContext ctx = (CharacterizationContext) object;
+			List<IItemPropertyDescriptor> characteristicBased = getCharacteristicBasedDescriptors(ctx);
+			return characteristicBased.stream().filter(ipd -> propertyId.equals(ipd.getId(object))).findAny()
+					.orElseGet(() -> super.getPropertyDescriptor(object, propertyId));
+		}
+		return super.getPropertyDescriptor(object, propertyId);
+	}
+
+	@Override
+	public Command createCommand(Object object, EditingDomain domain, Class<? extends Command> commandClass,
+			CommandParameter commandParameter) {
+		Command result = null;
+		if (commandParameter != null && commandParameter.owner instanceof CharacterizationContext) {
+			CharacterizationContext ctx = (CharacterizationContext) commandParameter.owner;
+
+			if (commandClass == SetCommand.class) {
+				result = createSetCommand(object, domain, commandClass, commandParameter, ctx);
+			} /*else if (commandClass == AddCommand.class) {
+				result = createAddCommand(object, domain, commandClass, commandParameter, ctx);
+			}*/
+		}
+		return result != null ? result : super.createCommand(object, domain, commandClass, commandParameter);
+	}
+
+	private Command createSetCommand(Object object, EditingDomain domain, Class<? extends Command> commandClass,
+			CommandParameter commandParameter, CharacterizationContext ctx) {
+		
+		if (commandParameter.getFeature() instanceof CharacteristicBasedVirtualStructuralFeature) {
+			CharacteristicBasedVirtualStructuralFeature feature = (CharacteristicBasedVirtualStructuralFeature) commandParameter.getFeature();
+
```

Comment:
Since the decorated code solely focuses on manifestations of characteristic based properties, the EAttribute case is not relevant. I adapted the code to ensure that it is only executed for virtual features and otherwise delegated to the decorated factory (there was in fact a bug, which would have prevented non-virtual features from being settable).